cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java

```diff
@@ -73,51 +75,64 @@
     long pollTimeMillis = cConf.getLong(Constants.RuntimeMonitor.POLL_TIME_MS);
     this.programRunsCache = CacheBuilder.newBuilder()
       .expireAfterWrite(pollTimeMillis, TimeUnit.MILLISECONDS)
-      .build(new CacheLoader<ProgramRunId, Boolean>() {
+      .build(new CacheLoader<ProgramRunId, Optional<ProgramRunInfo>>() {
+        /**
+         * For a programRunId this cache stores a {@link ProgramRunInfo} object only for valid programRunIds,
+         * i.e. this value will be Optional.empty() if the programRunId is invalid.
+         */
         @Override
-        public Boolean load(ProgramRunId programRunId) throws IOException, UnauthorizedException {
-          return isValid(programRunId);
+        public Optional<ProgramRunInfo> load(ProgramRunId programRunId)
+          throws IOException, UnauthorizedException {
+          return Optional.ofNullable(getRunRecordStatusForProgramsInNonEndState(programRunId));
```

Comment:
nit: may be just return optional from the method itself so you dont have to wrap it here with optional